File: 백준/Silver/14501. 퇴사/퇴사.py
n = int(input())
lst = [(0,0,0)]

for i in range(n):
  t, p = map(int,input().split())
  #t+i = 상담이 완료하는 날짜
  lst.append((i+1, p, t+i))

profit = [0]* (n+1) #i일의 최대 수익을 저장

max_p = 0
for day in range(1, n+1):#1,2,3,..n일에 최대 이익을 저장
  start, p, end = lst[day] 
  if end > n:
    continue

  pre = max(profit[start:end+1])
  if pre != 0: #lst[day] start와 이전에 상담한 날짜와 겹치는 경우
    
    max_p = max(pre, p + max(profit[:day]))
    profit[end] = max_p
  else:
    max_p = p + max(profit[:start])
    profit[end] = max_p #profit[i]의 이익과 lst[day]의 이익을 비교
  

print(max(profit))

File: 백준/Silver/14501. 퇴사/퇴사.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int n = Integer.parseInt(bf.readLine());
        int[][] schedule = new int[n][2];
       for (int i=0; i < n; i++) {
           int[] input = Arrays.stream(bf.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
           schedule[i][0] = input[0];
           schedule[i][1] = input[1];
       }

       int[] dp = new int[n];

       if (schedule[n-1][0] > 1) dp[n-1] = 0;
       else if (schedule[n-1][0] == 1) dp[n-1] = schedule[n-1][1];

       for (int i=n-2; i >= 0; i--) {
           if (i + schedule[i][0] > n) { //상담이 끝나는 날이 퇴사날 이후라면
               dp[i] = dp[i + 1];
               continue;
           }
           if (i + schedule[i][0] == n) {
               dp[i] = Math.max(schedule[i][1], dp[i+1]);
               continue;
           }
           dp[i] = Math.max(schedule[i][1] + dp[i+schedule[i][0]], dp[i+1]);
       }
        System.out.print(dp[0]);

        bf.close();
    }

}